## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [7]:
# What three item categories had the most entries?
# Assuming your DataFrame is named 'df'
top_categories = df['category'].value_counts().head(3)

# Display the result
print(top_categories)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [9]:
# For the category with the most entries, which subcategory had the most entries?
consumables_df = df[df["category"] == "consumables" ]
top_subcategory = consumables_df['subcategory'].value_counts().head(1)
print(top_subcategory)

subcategory
bathroom supplies    6424
Name: count, dtype: int64


In [10]:
# Which five clients had the most entries in the data?
top_client = df['client_id'].value_counts().head(5)
print(top_client)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [12]:
# Store the client ids of those top 5 clients in a list.
top_5_client_ids = df["client_id"].value_counts().head(5).index.tolist()
print(top_5_client_ids)

[33615, 66037, 46820, 38378, 24741]


In [16]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id = df["client_id"].value_counts().idxmax()
total_units_ordered = df[df['client_id'] == top_client_id]['qty'].sum()
print(total_units_ordered)

64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [18]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal'] = df['unit_price']* df['qty']
print(df[['unit_price','qty','line_subtotal']].head(2))

   unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95


In [21]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'].apply(lambda x: x * 7 if  x > 50 else x * 10)

print(df[['unit_price','unit_weight','qty','total_weight','shipping_price']].head())

   unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20


In [22]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['line_subtotal'] = df['unit_price'] * df['qty']
sales_tax_rate = 0.0925
df['line_price']= (df['line_subtotal']+ df['shipping_price']) * (1 + sales_tax_rate)
print(df[['line_subtotal','shipping_price','line_price']].head(3))

   line_subtotal  shipping_price     line_price
0      115164.00         5512.50  131839.076250
1         523.95          312.90     914.258625
2         527.28          458.64    1077.117600


In [34]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

df['line_cost'] = (df['unit_price'] * df['qty']) + df['shipping_price']
#print(df[['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id', 'unit_price', 'unit_cost', 'qty', 'line_subtotal', 'shipping_price', 'line_price', 'line_cost']].head(3))
df_subset = df.head(3)

df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

       line_subtotal  total_weight  shipping_price     line_price  line_cost  \
0          115164.00        787.50         5512.50  131839.076250  120676.50   
1             523.95         31.29          312.90     914.258625     836.85   
2             527.28         65.52          458.64    1077.117600     985.92   
3            1056.18         35.67          356.70    1543.571400    1412.88   
4            3902.00        928.60         6500.20   11364.403500   10402.20   
...              ...           ...             ...            ...        ...   
54634        2743.29         74.25          519.75    3564.871200    3263.04   
54635        9709.73        549.90         3849.30   14813.240275   13559.03   
54636       31188.50       1976.00        13832.00   49184.896250   45020.50   
54637         165.76       2020.48        14143.36   15632.713600   14309.12   
54638        3103.31       2134.17        14939.19   19711.431250   18042.50   

        line_profit  
0      11162.576250  
1         77.408625  
2         91.197600  
3        130.691400  
4        962.203500  
...             ...  
54634    301.831200  
54635   1254.210275  
54636   4164.396250  
54637   1323.593600

In [27]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['line_price']- df['line_cost']
print(df[['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id', 'unit_price', 'unit_cost', 'qty', 'line_subtotal', 'shipping_price', 'line_price', 'line_cost', 'line_profit']].head(3))
df.describe()

     first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  unit_price  \
0   harding.donald.7185@sullivan.com      58515   8953482     1096.80   
1  myers.t.6537@ferguson-johnson.net      37609   8069089       24.95   
2              swatson8146@payne.net      57113   1902144       13.52   

   unit_cost  qty  line_subtotal  shipping_price     line_price  line_cost  \
0     762.71  105      115164.00         5512.50  131839.076250  120676.50   
1      15.09   21         523.95          312.90     914.258625     836.85   
2       7.86   39         527.28          458.64    1077.117600     985.92   

    line_profit  
0  11162.576250  
1     77.408625  
2     91.197600  


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  line_subtotal  \
count  54639.000000  54639.000000  5.463900e+04  54639.000000   5.463900e+04   
mean      99.446073      5.004116  5.702646e+02      2.979667   7.710286e+04   
std      133.164267      5.326599  1.879552e+04      2.436320   2.881926e+06   
min        0.010000      0.000000  0.000000e+00      0.000000   0.000000e+00   
25%       14.840000      1.450000  3.200000e+01      1.000000   1.132950e+03   
50%       49.890000      3.240000  6.800000e+01      3.000000   4.629440e+03   
75%      125.570000      6.890000  1.700000e+02      5.000000   1.736913e+04   
max      846.270000     46.430000  3.958244e+06      9.000000   5.844743e+08   

       total_weight  shipping_price    line_price     line_cost   line_profit  
count  5.463900e+04    5.463900e+04  5.463900e+04  5.463900e+04  5.463900e+04  
mean   2.487416e+03    1.742510e+04  1.032718e+05  9.452796e+04  8.743836e+03  
std    5.559430e+04    3.891595e+05  3.286505e+06  3.008242e+06  2.782624e+05  
min    0.000000e+00    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
25%    7.033000e+01    4.966500e+02  2.887565e+03  2.643080e+03  2.444849e+02  
50%    2.254000e+02    1.577800e+03  8.396485e+03  7.685570e+03  7.109152e+02  
75%    7.184400e+02    5.029080e+03  2.645605e+04  2.421606e+04  2.239986e+03  
max    9.009056e+06    6.306339e+07  6.412625e+08  5.869680e+08  5.429454e+07

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [38]:
# Check your work using the totals above
order_totals = df.groupby('order_id')['line_price'].sum()
print(order_totals.loc[[2742071, 2173913, 6128929]])

order_id
2742071    152811.897075
2173913    162388.719300
6128929    923441.240450
Name: line_price, dtype: float64


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [41]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids. 
def client_totals(client_id, column):
    client_df = df.loc[df['client_id'] == client_id, column]
    return round(client_df.sum(), 2)
for client_id in top_5_client_ids:
    print(f"{client_id}: ${client_totals(client_id, 'line_price')}")

33615: $8377308.57
66037: $10259514.8
46820: $9743794.32
38378: $12906550.88
24741: $82268891.98


In [42]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_df = df.groupby('client_id').agg(
    qty=('qty', 'sum'),
    shipping_price=('shipping_price', 'sum'),
    line_price=('line_price', 'sum'),
    line_cost=('line_cost', 'sum'),
    line_profit=('line_profit', 'sum')
)
top_5_clients_summary = summary_df.sort_values(by='qty', ascending=False).head(5)
top_5_clients_summary

qty  shipping_price    line_price     line_cost   line_profit
client_id                                                                   
78965      4033302      5340756.88  6.559237e+08  6.003878e+08  5.553587e+07
86721      1010426     64335727.72  1.812663e+08  1.659188e+08  1.534749e+07
30143       846069      8981905.24  3.332885e+07  3.050696e+07  2.821893e+06
85931       644423     10122549.19  2.374170e+08  2.173153e+08  2.010167e+07
77354       506382     23814685.86  7.194265e+07  6.585140e+07  6.091254e+06

In [43]:
# Format the data and rename the columns to names suitable for presentation.


# Define the money columns. 
money_columns = ['shipping_price', 'line_price', 'line_cost', 'line_profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(x):
    return round(x / 1_000_000, 6)

# Apply the currency_format_millions function to only the money columns. 
top_5_clients_summary[money_columns] = top_5_clients_summary[money_columns].apply(currency_format_millions)


# Rename the columns to reflect the change in the money format. 
top_5_clients_summary.rename(columns={
    'qty': 'Units',
    'shipping_price': 'Shipping (millions)',
    'line_price': 'Total Revenue (millions)',
    'line_cost': 'Total Cost (millions)',
    'line_profit': 'Total Profit (millions)'
}, inplace=True)

top_5_clients_summary

Units  Shipping (millions)  Total Revenue (millions)  \
client_id                                                           
78965      4033302             5.340757                655.923670   
86721      1010426            64.335728                181.266297   
30143       846069             8.981905                 33.328850   
85931       644423            10.122549                237.417001   
77354       506382            23.814686                 71.942652   

           Total Cost (millions)  Total Profit (millions)  
client_id                                                  
78965                 600.387799                55.535871  
86721                 165.918807                15.347490  
30143                  30.506956                 2.821893  
85931                 217.315333                20.101668  
77354                  65.851398                 6.091254

In [45]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_df = top_5_clients_summary.sort_values(by='Total Profit (millions)', ascending=False)
sorted_df

Units  Shipping (millions)  Total Revenue (millions)  \
client_id                                                           
78965      4033302             5.340757                655.923670   
85931       644423            10.122549                237.417001   
86721      1010426            64.335728                181.266297   
77354       506382            23.814686                 71.942652   
30143       846069             8.981905                 33.328850   

           Total Cost (millions)  Total Profit (millions)  
client_id                                                  
78965                 600.387799                55.535871  
85931                 217.315333                20.101668  
86721                 165.918807                15.347490  
77354                  65.851398                 6.091254  
30143                  30.506956                 2.821893